### PR TITLE
[srpl] adding support for SRP replication

### PR DESCRIPTION
**Pull request by abtink**
This commit implements SRP replication (SRPL) which is a protocol for
ad-hoc replication of a DNS/SRP zone by multiple SRP servers. SRPL
eliminates the need for a single authoritative source of truth, e.g.,
a single primary DNS authoritative server. Each SRP replication
server maintains a copy of the SRP zone which is kept up to date.

As long as one SRPL peer remains online, SRP state is maintained even
if individual SRP replication peers (SRP servers) go offline. This
allows SRP servers and clients within the Thread network to use anycast
address model, which ensures that SRP clients do not see an apparent
change in servers and need to re-register when the server with which
they most recently registered with goes offline.

An SRPL peer discovers other peers by performing an ongoing DNS-SD
browse for the SRPL service name within the local browsing domain. An
SRPL DSO session is then established with each partner peer. This
session is first used for initial synchronization of all SRP related
info between the two peers, and then maintained during routine
operation for sending received SRP update messages from SRP clients
to the partner.

This commit adds `Srp::Srpl` class which contains the core
implementation of SRPL. It also updates `Srp::Server` to retain the
SRP update messages associated with different entries (which are used
by SRPL session during the initial synchronization phase). It adds a
set of public OT APIs to control SRP replication and get information
about list of partners and the state of their session. A set of CLI
commands is also added under `srp replication`. This commit also adds
a set of test-cases validating different behaviors of SRPL such as
discovering domain and partners, initial sync and routine operation
while adding, removing, and updating multiple services from different
clients.

-----

This PR contains commits from 
- https://github.com/openthread/openthread/pull/7258 

Please check and review only the last commit from this PR.

### CODE REVIEW

File: src/core/net/srp_replication.cpp [hunk 1]
```diff
@@ -0,0 +1,2486 @@
...
+
+    VerifyOrExit(allowsJoin);
+
+    switch (mState)
+    {
+    case kStateDiscovery:
+        if (mDomainName.IsNull())
+        {
+            // We accept any domain so adopt the domain name from partner.
+            SuccessOrExit(mDomainName.Set(static_cast<Heap::String &&>(domainName)));
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/srp_replication.cpp:269]: Yes. Exactly. 

At some point, we may add a helper method for this within OT core itself. In general, we cannot (and should not so to keep the code portable) use any STL functions/header within OT core (so cannot use `std::move()` here).

File: src/core/net/srp_replication.cpp [hunk 2]
```diff
@@ -0,0 +1,2486 @@
...
+{
+    AsCoreType(aInstance).Get<Srpl>().HandleDnssdBrowseResult(*static_cast<const Srpl::PartnerInfo *>(aPartnerInfo));
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// `Srpl`
+
+const char    Srpl::kTxtDataKeyDomain[]     = "domain";
+const char    Srpl::kTxtDataKeyAllowsJoin[] = "join";
+const char    Srpl::kTxtDataKeyPeerId[]     = "id";
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/srp_replication.cpp:68]: The PR followed draft 0 and some follow-up discussion we had from dnssd mailing list. I changed the key to "prec" now to match draft 1.

I noticed there are some (smaller) things/changes in draft 1 that are also not yet in this PR (e.g. dataset ID management which I will add it in follow up pushes). There are also some differences/ideas which may need to discuss further (like whether to include ID in SRPL Session request or not).

[the remark this answers is not in the document]
Comment (superwhd) [reply src/core/net/srp_replication.cpp]: I see. Thanks!

**Comment by abtink** [comment]
Closing this PR in favor of rebased and updated PR  https://github.com/openthread/openthread/pull/8403.

**Pull request closed without merging**